contrib/pg_tde/src/access/pg_tde_xlog_smgr.c

```diff
@@ -340,7 +422,10 @@ tdeheap_xlog_seg_read(int fd, void *buf, size_t count, off_t offset,
 	if (readsz <= 0)
 		return readsz;
 
-	TDEXLogCryptBuffer(buf, count, offset, tli, segno, segSize);
+	TDEXLogCryptBuffer(buf, buf, count, offset, tli, segno, segSize);
+
+	XLogSegNoOffsetToRecPtr(segno, offset + readsz, segSize, EncryptionState->LastReadOffset.lsn);
+	EncryptionState->LastReadOffset.tli = tli;
```

Comment:
That should be inside `if (RecoveryInProgress())`
otherwise replication (walsender) would constantly update it